src/python/WMCore/Database/MongoDB.py

```diff
@@ -0,0 +1,156 @@
+               the rest elements are considered to be indexes.
+               The indexes must beof type IndexModel. See pymongo documentation:
+
+               https://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.create_index
+        """
+        # DONE:
+        #     to create indexes on the proper fields
+
+        collName = coll[0]
+        collIndexes = list(coll[1:])
+        try:
+            self.client[db].create_collection(collName)
+        except errors.CollectionInvalid:
+            # this error is thrown in case of an already existing collection
+            pass
+        if collIndexes:
+            # DONE:
+            #    To implement a check for index types - should be either string or type IndexModel
+            #    see: https://api.mongodb.com/python/current/api/pymongo/operations.html#pymongo.operations.IndexModel
+            for index in collIndexes:
+                if not isinstance(index, IndexModel):
+                    msg = "ERR: Bad Index type for collection %s" % collName
+                    raise errors.InvalidName
+            try:
+                self.client[db][collName].create_indexes(collIndexes)
+            except Exception as ex:
+                msg = "ERR: Failed to create indexes on collection: %s\n%s" % (collName, str(ex))
+                self.logger.error(msg)
+                raise ex
+
+    def _dbTest(self, db):
+        # Test connection (from mongoDB documentation):
+        try:
+            # The 'ismaster' command is cheap and does not require auth.
+            self.client.admin.command('ismaster')
+        except errors.ConnectionFailure as ex:
+            msg = "Server not available: %s" % str(ex)
+            self.logger.error(msg)
+            raise ex
+
+        # Test for database existence
+        if db not in self.client.database_names():
+            msg = "ERR: Missing MongoDB databases: %s" % db
+            self.logger.error(msg)
+            raise errors.InvalidName
+
+    def _dbCreate(self, db):
+        # creating an empty collection in order to create the database
+        _initColl = self.client[db].create_collection('_initCollection')
+        _initColl.insert_one({})
+        # NOTE: never delete the _initCollection if you want the database to persist
+        # self.client[db].drop_collection('_initCollection')
+
+    def _dbConnect(self, db):
+        try:
+            setattr(self, db, self.client[db])
+            self._dbTest(db)
+        except errors.ConnectionFailure as ex:
+            msg = "ERR: Could not connect to MongoDB server for database: %s\n%s\n" % (db, str(ex))
+            msg += "Giving up Now."
+            self.logger.error(msg)
+            raise ex
+        except errors.InvalidName as ex:
+            msg = "ERR: Could not connect to a missing MongoDB databases: %s\n%s" % (db, str(ex))
+            self.logger.error(msg)
+            if self.create:
```

Comment:
If you are talking about having the database object creation fully outside the MSOutput module and put it somewhere on the top during the thread creation.... I can see the benefit from that... I agree. And even I already made some effort to move it as much as possible to the top of the initialization sequence (at the beginning it was inside the Producer function call), but I simply did not want to go out of the MSOutput.py boundaries for the time being. I would say, lets go with that as it is right now (we can berry a repetition of object recreation once per pooling cycle) and we create a separate issue and pull this piece out of the module next month, because it will need additional testing.  

In case I misunderstood you please cerrect  me :)

Reply:
Done: See https://github.com/dmwm/WMCore/issues/9711